File: engine/negative_controls/ssrf.py
"""Negative control: "ssrf" keyword family (Python).

Outbound requests use fixed literal URLs or an allowlist lookup whose
result is a bare identifier. No f-string URL, no concatenation, no
request-derived URL argument — the unsafe-shape pattern must not match.
"""

import requests

ALLOWED_ENDPOINTS = {
    "status": "https://api.example.com/status",
    "health": "https://api.example.com/health",
}


def fetch_status():
    return requests.get("https://api.example.com/status", timeout=5)


def fetch_named(name):
    if name not in ALLOWED_ENDPOINTS:
        raise ValueError("unknown endpoint")
    endpoint = ALLOWED_ENDPOINTS[name]
    return requests.get(endpoint, timeout=5)

File: engine/negative_controls/ssrf.c
/* Negative control: "ssrf" keyword family (C).
 *
 * libcurl fetch with a compile-time constant URL — no user data
 * reaches CURLOPT_URL. The unsafe-shape pattern (concatenation /
 * argv / buffer flowing into CURLOPT_URL) must not match.
 */
#include <curl/curl.h>

#define STATUS_URL "https://api.example.com/status"

int fetch_status(void)
{
    CURL *curl = curl_easy_init();
    CURLcode rc;

    if (!curl) {
        return -1;
    }
    curl_easy_setopt(curl, CURLOPT_URL, STATUS_URL);
    rc = curl_easy_perform(curl);
    curl_easy_cleanup(curl);
    return rc == CURLE_OK ? 0 : -1;
}
